### ISSUE TITLE
getsource fails for functions defined inside an `exec`

### ISSUE THREAD

**Issue by mmckerns** [uqfoundation/dill#145]
`dill.source.getsource` will fail for ```exec('f = lambda x:x')```

**Comment by mmckerns** [comment]
Also see: https://stackoverflow.com/questions/35320721/dilling-error-its-not-found-in-builtins/

Essentially, this will fail:
```
>>> import dill
>>> local_env = {}
>>> global_env = {}
>>> x = "class Human(object):\n  pass"
>>> exec(x, global_env, local_env)
>>> local_env
{'Human': <class 'Human'>}
>>> dill.dumps(local_env['Human'])
```

**Comment by briantom-jetbrains** [comment]
```
import dill
import builtins
x = """
def test(x, y):
    return x + y
"""
env = {}
env['__builtins__'] = builtins
local_env = Environment()
exec(x, env, local_env)
exec("__builtins__.{0} = {0}".format('test'), env, local_env)
test = dill.loads(dill.dumps(local_env['test']))
```

This could be used as a workaround for now if necessary.

**Comment by mmckerns** [comment]
@briantom-jetbrains: yeah, that or some variant could work.  `__builtins__` should probably get cleaned up afterward.  And I'm assuming `Environment()` is just some generic expression of `locals()` or whatever, right?

**Comment by briantom-jetbrains** [comment]
ah yes. That was not clear. Environment is a generic instance of an environment. I was using dill for copying purposes and this is the full snippet.

```
def serial_copy(original_object):
    try:
        key = original_object.__name__
        if re.match("[_A-Za-z][_a-zA-Z0-9]*$", key) is None:
            key = "dummy_key"
    except AttributeError:
        key = 'dummy_key'
    dummy_dic = {'__builtins__': builtins}
    dict.__setitem__(dummy_dic, key, original_object)
    exec("__builtins__.{0} = {0}".format(key), dummy_dic)
    x = dill.dumps(original_object)
    y = dill.loads(x)
    exec("del __builtins__.{0}".format(key), dummy_dic)
    return y
```

**Comment by mmckerns** [comment]
@briantom-jetbrains: Ok, thanks -- that looks like the start of a good workaround.  For it to go into `dill`, it would need to her refactored into a `dumps` with no side-effects, and a `load`.  A worry would be the potential to overwrite an existing key in `builtins`, so that would need to be addressed. Additionally, where would it go?  You can't assume you know the object was created with an `exec`, and probably shouldn't be messing with `builtins` if it's not needed.  I'd expect it's have to be tried when the function/class fails to be located.  I don't like the use of `re.match`, primarily because this will be slow -- however, it's a good start.

When the above are resolved, then I think it would definitely be worthwhile to add to `dill`.

**Comment by briantom** [comment]
I am trying to come up with a robust fix for this right now and I was wondering what exactly you meant by "parses the builtins buffer for class definitions". Maybe you could point to where that is being done in the source code. Thank you.

**Comment by mmckerns** [comment]
Cool, I'll be interested to see that. 

What I meant by "parses the buffer" is that python holds a memory buffer for what gets typed in the interpreter… and that can be accessed, and parsed… and `dill` does that.  See almost everything in `dill.source`  (https://github.com/uqfoundation/dill/blob/master/dill/source.py)… for example, `dill.source.findsource`.  For objects defined in `__main__`, `__main__` is treated much like it were a module, and basically, this "file" is also read from the buffer… this is also important when searching for the global object references for functions… as is done in `dill.dill.save_function` (see here: https://github.com/uqfoundation/dill/blob/master/dill/dill.py#L776, and subsequently, `dill.detect.globalvars`.
